client/csi_endpoint.go

```diff
@@ -171,6 +171,137 @@ func (c *CSI) ControllerDetachVolume(req *structs.ClientCSIControllerDetachVolum
 	return nil
 }
 
+func (c *CSI) ControllerCreateVolume(req *structs.ClientCSIControllerCreateVolumeRequest, resp *structs.ClientCSIControllerCreateVolumeResponse) error {
+	defer metrics.MeasureSince([]string{"client", "csi_controller", "create_volume"}, time.Now())
+
+	plugin, err := c.findControllerPlugin(req.PluginID)
+	if err != nil {
+		// the server's view of the plugin health is stale, so let it know it
+		// should retry with another controller instance
+		return fmt.Errorf("%w: %v", nstructs.ErrCSIClientRPCRetryable, err)
+	}
+	defer plugin.Close()
+
+	csiReq, err := req.ToCSIRequest()
+	if err != nil {
+		return err
+	}
+
+	ctx, cancelFn := c.requestContext()
+	defer cancelFn()
+
+	// CSI ControllerCreateVolume errors for timeout, codes.Unavailable and
+	// codes.ResourceExhausted are retried; all other errors are fatal.
+	cresp, err := plugin.ControllerCreateVolume(ctx, csiReq,
+		grpc_retry.WithPerRetryTimeout(CSIPluginRequestTimeout),
+		grpc_retry.WithMax(3),
+		grpc_retry.WithBackoff(grpc_retry.BackoffExponential(100*time.Millisecond)))
+	if err != nil {
+		return err
```

Comment:
The error gets logged by the client's `handleNomadConn` method. These are 1:1 with the plugin method but wrapping it could probably make the path through the code more clear.

Reply:
For context, I've been bitten before with very non-descriptive error messages like "transport closing" that could have been generated from multiple call sites within the endpoint handling and it was hard to know exactly where the error got generated.

Reply:
Wrapped all these client errors in 0a39122